### PR TITLE
Add tests for base Editor object

### LINKED ISSUE #608: Editor._sync_values is broken
This line https://github.com/enthought/traitsui/blob/master/traitsui/editor.py#L450 gives
```
ValueError: too many values to unpack (expected 2)
```

Fortunately nothing is using this at the moment.

### PR DESCRIPTION
**Pull request by corranwebster**
TraitsUI needs more tests of core functionality.  This is a reasonably comprehensive set of tests for the base `Editor` object.

Incidentally fixes #608.

### CODE REVIEW

**Comment by corranwebster** [comment]
@mdickinson if you have time, could you review (this is hopefully an easy one)

File: traitsui/editor.py [hunk 1]
```diff
@@ -447,7 +447,7 @@ def _sync_values(self):
             value of corresponding factory traits.
         """
         factory = self.factory
-        for name, trait in factory.traits(sync_value=not_none):
+        for name, trait in factory.traits(sync_value=not_none).items():
```
Comment (mdickinson) [inline traitsui/editor.py]: Gah. How on earth was this working before?

File: traitsui/tests/test_editor.py [hunk 2]
```diff
@@ -0,0 +1,733 @@
...
+from traits.trait_base import xgetattr
+from traits.testing.unittest_tools import UnittestTools
+
+from traitsui.context_value import ContextValue, CVFloat, CVInt
+from traitsui.editor import Editor
+from traitsui.editor_factory import EditorFactory
+from traitsui.handler import default_handler
+from traitsui.ui import UI
+
+from ._tools import is_current_backend_null
```
Comment (mdickinson) [inline traitsui/tests/test_editor.py]: Can `_tools` be promoted to public to avoid the private import? (Not in this PR.)

File: traitsui/tests/test_editor.py [hunk 3]
```diff
@@ -0,0 +1,733 @@
...
+                    self.value = new
+                finally:
+                    self._no_update = False
+
+    def set_focus(self, parent):
+        pass
+
+
+class UserObject(HasTraits):
+
```
Comment (mdickinson) [inline traitsui/tests/test_editor.py]: Would be nice to have short docstrings on these classes. I could imagine re-using this code (via copy and paste, rather than import) elsewhere.

File: traitsui/tests/test_editor.py [hunk 4]
```diff
@@ -0,0 +1,733 @@
...
+    user_auxiliary = Any(10)
+
+    #: An list user value
+    user_list = List(['one', 'two', 'three'])
+
+    #: An event user value
+    user_event = Event
+
+
+@unittest.skipIf(is_current_backend_null(),
```
Comment (mdickinson) [inline traitsui/tests/test_editor.py]: Might be worth having a "requires_non_null_backend" decorator in `_tools` for this sort of thing. (Naming to be bikeshedded, and not as part of this PR.)

File: traitsui/tests/test_editor.py [hunk 5]
```diff
@@ -0,0 +1,733 @@
...
+            ui=ui,
+            object_name=object_name,
+            name=name,
+            factory=factory,
+            object=user_object,
+        )
+        return editor
+
+    def test_lifecycle(self):
+        editor = self.create_editor()
```
Comment (mdickinson) [inline traitsui/tests/test_editor.py]: Should we have some kind of try/finally or context-manager setup to make sure that editors always get disposed in these tests, even if one of the test assertions has failed?

File: traitsui/editor.py (hunk 1 above)
Comment (corranwebster) [reply traitsui/editor.py]: It wasn't.  Nothing was using it.

File: traitsui/tests/test_editor.py (hunk 5 above)
Comment (corranwebster) [reply traitsui/tests/test_editor.py]: In this particular test suite it doesn't matter because there is no actual toolkit control being created - everything (including the `FakeControl`) is pure Python.

File: traitsui/tests/test_editor.py (hunk 4 above)
Comment (corranwebster) [reply traitsui/tests/test_editor.py]: There is in fact such a decorator, but it only works with methods and not with classes at the moment :(

File: traitsui/tests/test_editor.py (hunk 2 above)
Comment (corranwebster) [reply traitsui/tests/test_editor.py]: Yes, I'd like that - in fact it probably belongs in a traitsui.testing module or some such.

File: traitsui/tests/test_editor.py (hunk 3 above)
Comment (corranwebster) [reply traitsui/tests/test_editor.py]: Done.

**Comment by corranwebster** [comment]
Arg - tests don't run stand-alone because there are a couple of buried `dispatch='ui'` tucked away in the code (that's why the `null` tests fail).  Needs to be refactored to use the event loop when testing setting traits on the model object (this is why tests were failing on the null backend - no event loop, so no `ui` dispatch).
